### ISSUE TITLE
Add command to restart RLS and rebuild all projects

### ISSUE THREAD

**Issue by norru** [eclipse/corrosion#165]
There are many instance in which changes to the project outside of the IDE makes RLS in the IDE go out of sync.

In other cases the RLS (or is it LSP?) simply stalls and get stuck performing indexing or build operations.

Closing and reopening all editor tabs does not reliably restart the RLS.

In those cases the only way to force a reload/rebuild of the project (and perhaps a restart of RLS) is to close and reopen the IDE.

There should be an explicit command which forces a restart of all instances of RLS and triggers a reload/reindex of all affected projects in the workspace, also triggering a build if required.

**Comment by norru** [comment]
https://github.com/eclipse/corrosion/issues/164

**Comment by mickaelistria** [comment]
I'm not really motivated in such workarounds instead of fixing the root cause. The sync issue is what needs to be fixed more than the restart RLS.
Not that the Eclipse IDE already has in its preferences for the workspace an option.to refresh its content automatically. That might be helpful.in this cas,but it can a relatively expensive setting in term of CPU.

**Comment by norru** [comment]
TL;DR This feature request is a stopgap.

I agree that the fixing the root cause is the best way to proceed. However, bug fixing in this project has not a stellar track record in terms of speed, mostly caused by the tight dependencies, so my fear is that the "shut down and start again" workaround will keep tormenting everybody for months or years to come.

Bear in mind that it's not always the same bug/condition causing the hang. The restart is just a "get out of jail" card which is needed exactly to cope with new/unfixed bugs while they are being addressed.

Manually refreshing the project doesn't fix the problem. I still have to restart the IDE after a manual refresh as RLS doesn't pick them up, so auto refresh probably won't make any difference. Will try anyway.

**Comment by mickaelistria** [comment]
> I agree that the fixing the root cause is the best way to proceed. However, bug fixing in this project has not a stellar track record in terms of speed, mostly caused by the tight dependencies, so my fear is that the "shut down and start again" workaround will keep tormenting everybody for months or years to come.

It's up to the code contributors to decide their priority.
I personally won't spend any second implementing a workaround instead of trying to fix the bug. I'll try to fix some bugs on that stack when I have some availability. I also suggest any contributor who has time to dedicated to Corrosion to do the same: prioritize real bug fixes on the right layer over workarounds. It's the only way for a project to be sustainable.

**Comment by norru** [comment]
That's fair enough. 

However, what are the options here? Should I request this feature to LSP4E instead? Or ask RLS team to make sure the RLS never hangs? Or just put up and shut up and restart the IDE when this happens (many, many times a day?). I'd be tempted to go back to CLion but the idea of that horrible UI forces me to stick around ;)

I would love to contribute and just put the feature in myself, but I have limited spare time so it's either doing this, or the Rust projects I am actually using the IDE for...

**Comment by norru** [comment]
Also, by the way, thanks **a lot** for your help and contributions on this, much appreciated :)

**Comment by mickaelistria** [comment]
> Should I request this feature to LSP4E instead? 

FWIW, I'm also project lead for LSP4E, so it's not very likely that my opinion differs in LSP4E and that I'll prioritize the workaround implementation (it's not a feature by the way) over the bugfixes ;)
But there are some other contributors in LSP4E who may want to implement it if you ask.

> Or ask RLS team to make sure the RLS never hangs?

If this is a bug in RLS, a real hang, then yes, a dedicated bug should be opened to RLS to make it more robust.

> However, what are the options here?
> ...
> I would love to contribute and just put the feature in myself, but I have limited spare time so it's either doing this, or the Rust projects I am actually using the IDE for...

I think giving details in the bugs like you do already is a lot. As an addition, when you see a freeze/hang, then you can use `jstack` to get the current stack state of the IDE and put it on the ticket. It could help to identify possible deadlocks and possible remediation.
If at some point, you feel like you could start debugging Corrosion or LSP4E directly and submit fixes, that would be great; but I understand it's not always possible, and good quality bug reports as you submit are already great contributions.

The main issue is that contributors to Corrosion and LSP4E are currently not numerous and available enough to react promptly to your reports. Most of the few of us are also working on other projects that have higher priority, and Corrosion (and somehow even LSP4E) is currently maintained in a best effort possibility, which leads to slow reaction time and slow fixes.
I wished I could just say "I'm having a look at it right now and hope this can be fixed by the end of the week", but that would be a lie; in all honesty, I have to say "I'll have a look when I can, and it's not before next week, and I'm not even sure I'll find a proper fix" ;)

**Comment by norru** [comment]
I've just got a similar problem
- "building" stuck at 0%
- manually interrupted the build
- RLS dead
- this in the Error log:

```
java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.eclipse.lsp4e.operations.hover.LSBasedHover.getHoverRegion(LSBasedHover.java:260)
	at org.eclipse.ui.internal.genericeditor.hover.CompositeTextHover.getHoverRegion(CompositeTextHover.java:101)
	at org.eclipse.jface.text.TextViewerHoverManager.computeInformation(TextViewerHoverManager.java:142)
	at org.eclipse.jface.text.AbstractInformationControlManager.doShowInformation(AbstractInformationControlManager.java:1134)
	at org.eclipse.jface.text.AbstractHoverInformationControlManager$MouseTracker.mouseHover(AbstractHoverInformationControlManager.java:511)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:212)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5797)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1374)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5051)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4583)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:156)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:628)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:563)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:151)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:199)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:391)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:246)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:595)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1501)
```

**Comment by norru** [comment]
Another instance, here's the jstack output
[jstack_building.txt](https://github.com/eclipse/corrosion/files/2572738/jstack_building.txt)

**Comment by mickaelistria** [comment]
Won't fix: we don't have enough resources to waste them in implementing workarounds, and restarting the IDE is already a decent workaround. Effort have to be placed on the root cause of the issues, wherever they are, than in local workarounds that do not cascade to value in other affected components.

**Comment by norru** [comment]
I disagree with the resolution. I am still restarting the IDE many many times a day and it's a very big nuisance.

**Comment by mickaelistria** [comment]
Let's find the root cause and fix them then.

**Comment by norru** [comment]
I've posted separate bugs for all known cases. It's not one root cause unless you want to call "general flakyness" a root cause.

**Comment by mickaelistria** [comment]
I (and my teammates) won't spend any second on a workaround when I already struggle to find time doing proper fixes.
If you want it, you'll have to implement it ;)

**Comment by norru** [comment]
FYI https://github.com/microsoft/vscode/issues/76405

**Comment by norru** [comment]
Another issue which requires restart: https://github.com/eclipse/corrosion/issues/261

**Comment by norru** [comment]
I'd reopen this in case someone else wants to volunteer :)

**Comment by mickaelistria** [comment]
> Another issue which requires restart
>
Again, it doesn't require restart, it requires a fix.

**Comment by norru** [comment]
> Again, it doesn't require restart, it requires a fix.

I do agree, except if I want to actually do any work while the issues are not yet fixed I do require a restart.

**Comment by mickaelistria** [comment]
Let's close this one and focus on fixing bugs that may make restart useful.
